enhancements/oc/image-content-source-policy-awareness.md

```diff
@@ -0,0 +1,87 @@
+---
+title: image-content-source-policy-awareness
+authors:
+  - "@sallyom"
+reviewers:
+  - "@smarterclayton"
+  - "@soltysh"
+  - "@wking"
+approvers:
+  - "@smarterclayton"
+creation-date: 2020-05-19
+last-updated: 2020-05-19
+status: provisional
+---
+
+# ImageContentSourcePolicy (ICSP) Awareness
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  
+
+There have been several bugs opened around the experience of a 
+user in a disconnected environment using `oc adm release` commands.  If
+using a mirrored image and the mirrored source registry is connected, 
+the following commands do not succeed when in a disconnected environment:
+
+```console
+$ oc adm release extract --tools aprivateregistry/repo/name:tag
+$ oc adm release mirror aprivateregistry/repo/name:tag --to someregistry/repo/name
+$ oc adm must-gather
+```
+
+This is because the mirrored image tags (the individual component images from a payload)
+retain references to the mirrored registry, usually something like 
+`quay.io/openshift-release-dev/ocp-v4.0-art-dev`.  
+
+There needs to be logic in `oc` to look for `ImageContentSourcePolicy` either from a file or from a cluster.
```

Comment:
I don't think a file is required, although might be nice.

All `oc image` and `oc adm release` commands need to support this.  However, `oc image` commands are not allowed to talk to API servers (they are commands that are cluster agnostic).

So this solution needs to implement something that works for `oc image` without the logic described about looking up `ICSP` from server.

Reply:
Any flags added for `oc image mirror` can we also add to `oc adm catalog mirror`?

Reply:
Yes, we'll add any function across the board for all oc adm cmds to handle ICS information. 
The file will be created during the mirror, so oc can then use the info from the ICSP file for other cmds, if it exists, like so: https://github.com/openshift/oc/pull/439  that the `oc adm catalog mirror` already writes.